File: sdk/contextos.py
"""Dependency-free client for the trusted localhost owner API (Python 3.10+)."""
import json
from urllib.error import HTTPError
from urllib.parse import quote, urlencode
from urllib.request import Request, urlopen


class ContextOS:
    def __init__(self, url="http://127.0.0.1:3000"):
        self.url = url.rstrip("/")

    def request(self, path, method="GET", body=None):
        data = json.dumps(body).encode() if body is not None else None
        request = Request(self.url + "/api" + path, data=data, method=method,
                          headers={"Content-Type": "application/json"} if data else {})
        try:
            with urlopen(request, timeout=30) as response:
                return json.load(response)
        except HTTPError as error:
            try:
                message = json.loads(error.read()).get("error", str(error))
            except (ValueError, AttributeError):
                message = str(error)
            raise RuntimeError(message) from error

    def recall(self, query=""):
        return self.request("/memories?" + urlencode({"q": query}))

    def get(self, memory_id):
        return self.request("/memories/" + quote(memory_id, safe=""))

    def remember(self, memory):
        memory_id = memory.get("id")
        return self.request("/memories" + ("/" + quote(memory_id, safe="") if memory_id else ""),
                            "PUT" if memory_id else "POST", memory)

    def history(self, query=""):
        return self.request("/events?" + urlencode({"q": query}))

    def handoff(self, packet):
        return self.request("/handoffs", "POST", packet)

    def export(self):
        return self.request("/export")

    def import_archive(self, archive):
        return self.request("/import", "POST", archive)

    def tool(self, name, arguments=None, agent="opencode"):
        return self.request("/tools/call", "POST", {"name": name, "arguments": arguments or {}, "agent": agent})

File: sdk/contextos.mjs
/** Small, dependency-free client for the trusted localhost owner API. */
export class ContextOS {
  constructor(url = 'http://127.0.0.1:3000') { this.url = url.replace(/\/$/, ''); }
  async request(path, method = 'GET', body) {
    const response = await fetch(`${this.url}/api${path}`, { method, headers: body === undefined ? {} : { 'Content-Type': 'application/json' }, body: body === undefined ? undefined : JSON.stringify(body) });
    const data = await response.json();
    if (!response.ok) throw new Error(data.error || `ContextOS returned ${response.status}`);
    return data;
  }
  recall(query = '') { return this.request(`/memories?q=${encodeURIComponent(query)}`); }
  get(id) { return this.request(`/memories/${encodeURIComponent(id)}`); }
  remember(memory) { return this.request(memory.id ? `/memories/${encodeURIComponent(memory.id)}` : '/memories', memory.id ? 'PUT' : 'POST', memory); }
  history(query = '') { return this.request(`/events?q=${encodeURIComponent(query)}`); }
  handoff(packet) { return this.request('/handoffs', 'POST', packet); }
  export() { return this.request('/export'); }
  import(archive) { return this.request('/import', 'POST', archive); }
  tool(name, args = {}, agent = 'opencode') { return this.request('/tools/call', 'POST', { name, arguments: args, agent }); }
}
